### DISCUSSION TITLE
Which image is used for the list

### DISCUSSION THREAD

**Discussion by jstride** [jocmp/capyreader#562] [Q&A] [answered]
How does capy reader choose the image to use in the list view? I have items with no images but the RSS has an image.

**Comment by jstride** [comment]
You can see here that in freshrss items have a thumbnail but not all corresponding items in Capy Reader

![Screenshot_20241127-053342.png](https://github.com/user-attachments/assets/f12b0164-ddb3-4542-8eab-731419f7386c)

![IMG_20241127_053320_726.jpg](https://github.com/user-attachments/assets/2b5642e6-ca17-43ec-b009-75d8fe96d6e4)

**Reply by jocmp** [reply]
Hey @jstride - Capy doesn't do any client-side logic to grab images for FreshRSS. Those are taken directly from the `enclosure` list that's returned from FreshRSS ([capy source](https://github.com/jocmp/capyreader/blob/31549bd1ea9c4315298f9ff5d7ee8ccf207cc04b/readerclient/src/main/java/com/jocmp/readerclient/Item.kt#L47-L48)) ([FreshRSS source](https://github.com/FreshRSS/FreshRSS/blob/c1c2dd83e401b84b74a77794344dcf6d6e56d886/app/Models/Entry.php#L321-L340)).

So it's up to whether the FreshRSS feed returns enclosures. I added the New Yorker feed to my FreshRSS instance. It doesn't include an enclosure but does include images in the content body.

<details>
<summary>New Yorker JSON</summary>
<pre lang="jsonc">
{
  "id": "user/-/state/com.google/reading-list",
  "updated": 1732673243,
  "items": [
    {
      "id": "tag:google.com,2005:reader/item/000627db6a60641e",
      "crawlTimeMsec": "1732673196287",
      "timestampUsec": "1732673196287006",
      "published": 1732652880,
      "title": "Sarah McBride Wasn’t Looking for a Fight on Trans Rights",
      "canonical": [
        {
          "href": "https://www.newyorker.com/podcast/the-new-yorker-radio-hour/sarah-mcbride-wasnt-looking-for-a-fight-on-trans-rights"
        }
      ],
      "alternate": [
        {
          "href": "https://www.newyorker.com/podcast/the-new-yorker-radio-hour/sarah-mcbride-wasnt-looking-for-a-fight-on-trans-rights"
        }
      ],
      "categories": [
        "user/-/state/com.google/reading-list",
        "user/-/label/Uncategorized",
        "user/-/state/com.google/read",
        "Podcast / The New Yorker Radio Hour"
      ],
      "origin": {
        "streamId": "feed/44",
        "htmlUrl": "https://www.newyorker.com/",
        "title": "The New Yorker"
      },
      "summary": {
        "content": "The first transgender person elected to Congress discusses how to respond to a bathroom bill and transphobic attacks from other House members, including Speaker Mike Johnson.<figure class=\"enclosure\">\n\t<p class=\"enclosure-content\">\n\t\t<img class=\"enclosure-thumbnail\" src=\"https://media.newyorker.com/photos/674607a9f5f367966198ec9b/master/pass/Radio_Hour_Sarah_Mcbride_FINAL_11_25_2024.jpg\" alt="" />\n\t</p>\n</figure>"
      }
    }
  ]
}
</pre>
</details>

Contrast that to the NASA Image of the Day which does return an enclosure.

<details>
<summary>NASA Image of the Day JSON</summary>
<pre lang="jsonc">
{
  "id": "user/-/state/com.google/reading-list",
  "updated": 1732673376,
  "items": [
    {
      "id": "tag:google.com,2005:reader/item/000627db746b1697",
      "crawlTimeMsec": "1732673364760",
      "timestampUsec": "1732673364760215",
      "published": 1732652220,
      "title": "Selfie Time with Astronaut Victor Glover",
      "canonical": [
        {
          "href": "https://www.nasa.gov/image-detail/afs-8-101-919/"
        }
      ],
      "alternate": [
        {
          "href": "https://www.nasa.gov/image-detail/afs-8-101-919/"
        }
      ],
      "categories": [
        "user/-/state/com.google/reading-list",
        "user/-/label/Uncategorized"
      ],
      "origin": {
        "streamId": "feed/46",
        "htmlUrl": "https://www.nasa.gov/image-of-the-day/",
        "title": "NASA Image of the Day"
      },
      "summary": {
        "content": "NASA astronaut Victor Glover (right) takes photos during a visit on Friday, Nov. 8, 2024, with employees at NASA’s Kennedy Space Center in Florida. The employees support the agency’s Exploration Ground Systems Program to help NASA send astronauts, including Glover, to the Moon and back through the Artemis II launch.\n<figure class=\"enclosure\"><p class=\"enclosure-content\"><img src=\"https://www.nasa.gov/wp-content/uploads/2024/11/ksc-20241108-ph-jbs01-0065orig.jpg\" alt=\"\" title=\"\" /></p></figure>\n"
      },
      "enclosure": [
        {
          "href": "https://www.nasa.gov/wp-content/uploads/2024/11/ksc-20241108-ph-jbs01-0065orig.jpg",
          "type": "image/jpeg",
          "length": 20095711
        }
      ]
    }
  ]
}
</pre>
</details>

**Reply by jstride** [reply]
Thanks @jocmp that absolutely makes sense, so for the following RSS which includes a media:thumbnail - would Fresh RSS need to parse that into the GReader API response?

https://feeds.businessinsider.com/custom/all

Also, the following appears to have an image but that doesn't appear via Fresh RSS
https://www.theguardian.com/au/business/rss

**Reply by jocmp** [reply]
Yep! The media:thumbnail would work though it's not returned as a part of the enclosure at the moment. There's also a lead image in the Business Insider articles (sample below). It may be a missing case in the regex ([FreshRSS source code](https://github.com/FreshRSS/FreshRSS/blob/256dcc21bb222184d5e917ea57cec334b74b96f4/app/Models/Entry.php#L289)) doesn't pick up on the escaped `&lt;` character.

```xml
  <entry>
    <title>I ordered the same turkey sandwich from Subway and Potbelly Sandwich Shop, and I preferred the cheaper option</title>
    <content type="html">&lt;figure&gt;&lt;img src=&#34;https://i.insider.com/6745ff20ede4eeae3929a5ed?format=jpeg&#34; height=&#34;1500&#34; width=&#34;2000&#34; charset=&#34;&#34; alt=&#34;subway and potbelly sandwich shop turkey sandwiches with logos&#34;/&gt;&lt;figcaption&gt;I tried turkey sandwiches from Subway and Potbelly Sandwich Shop.&lt;p class=&#34;copyright&#34;&gt;Erin McDowell; Birch Photographer/Shutterstock; Alexander Oganezov
```

For The Guardian feed, these images return 401 failures for me ([example](https://i.guim.co.uk/img/media/e42757d4b123d0f6cfa57cd5207e0d412d1a4da1/0_223_4288_2573/master/4288.jpg?width=460&amp;quality=85&amp;auto=format&amp;fit=max&amp;s=226b735082a710447fa0348775ad6beb)). I can add better handling to omit error'd images from the feed but there might be something FreshRSS can do too to prevent sending unavailable images.

**Comment by jocmp** [comment, marked as answer]
I have a fix for this that went out ~2 weeks ago as of 2025.01.1090-dev. FreshRSS images are now pulled from the summary content if available and then enclosures as a fallback.

- https://github.com/jocmp/capyreader/issues/635
